### PR TITLE
Add changes for pushforward trick

### LINKED ISSUE #1740: Run pushforward trick experiments
### Describe the task. It can be a feature, documentation, etc.

The idea is to apply loss function on the last forecast step rather applying to all intermediate steps
The experiments will be conducted using both [fe-param-search](https://github.com/ecmwf/WeatherGenerator/tree/mk/develop/fe-param-search)  branch and `develop` branch.

### Hedgedoc URL, if you are keeping notes, plots, logs in hedgedoc.

_No response_

### Area

- [ ] datasets, data readers, data preparation and transfer
- [ ] model
- [x] science
- [ ] infrastructure and engineering
- [ ] evaluation, export and visualization
- [ ] documentation

### PR DESCRIPTION
**Pull request by SavvasMel**
## Description

This PR adds the necessary changes for the pushforward trick.

## Issue Number

Fixes https://github.com/ecmwf/WeatherGenerator/issues/1740

See Hedgedoc: https://gitlab.jsc.fz-juelich.de/hedgedoc/bGdOXZgDTTyX074RRvliWg?both

## Checklist before asking for review

-   [X] I have performed a self-review of my code
-   [ ] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [X] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [ ] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

### CODE REVIEW

File: src/weathergen/model/model_interface.py [hunk 1]
```diff
@@ -111,7 +111,10 @@ def init_model_and_shard(
 
         for module in model.forecast_engine.fe_blocks.modules():
             if isinstance(module, modules_to_shard):
-                fully_shard(module, **fsdp_kwargs)
+                # reshard_after_forward=False keeps FE parameters unsharded
+                # during the multi-step rollout loop.
+                # Needed for pushforward trick.
+                fully_shard(module, reshard_after_forward=False, **fsdp_kwargs)
```
Comment (clessig) [inline src/weathergen/model/model_interface.py:113]: @sophie-xhonneux : is this maybe related to the problem we are seeing with the EMATeacher where we need to reshard?

File: src/weathergen/model/model.py [hunk 2]
```diff
@@ -695,15 +695,34 @@ def forward(self, model_params: ModelParams, batch: ModelBatch) -> ModelOutput:
...
+                    tokens = self.forecast_engine(tokens, step, coords=model_params.rope_coords)
+                    output = self.predict_latent(model_params, step, tokens, batch, output)
+                    output = self.predict_decoders(model_params, step, tokens, batch, output)
+                else:
+                    # Pushforward mode: only add empty predictions (no grad, no engine call)
+                    with torch.no_grad():
+                        tokens = self.forecast_engine(tokens, step, model_params.rope_coords)
+
+                    # Add empty predictions for all streams (vectorized / batched if possible)
+                    for stream_name in self.stream_names:
```
Comment (clessig) [inline src/weathergen/model/model.py]: We can avoid this when we create `output` with the correct length. I thought we do this anyway, and hence the `step` argument.

File: src/weathergen/model/model.py [hunk 3]
```diff
@@ -695,15 +695,34 @@ def forward(self, model_params: ModelParams, batch: ModelBatch) -> ModelOutput:
...
-            # latent predictions (raw and with SSL heads)
-            output = self.predict_latent(model_params, step, tokens, batch, output)
+                needs_full_prediction = (
+                    not pushforward_trick
+                    or not self.training
+                    or step == max(batch.get_output_idxs())
+                )
+
+                if needs_full_prediction:
+                    tokens = self.forecast_engine(tokens, step, coords=model_params.rope_coords)
```
Comment (clessig) [inline src/weathergen/model/model.py]: Remove `coords=` for consistency

File: src/weathergen/model/model.py [hunk 4]
```diff
@@ -695,15 +695,34 @@ def forward(self, model_params: ModelParams, batch: ModelBatch) -> ModelOutput:
...
-            output = self.predict_decoders(model_params, step, tokens, batch, output)
-            # latent predictions (raw and with SSL heads)
-            output = self.predict_latent(model_params, step, tokens, batch, output)
+                needs_full_prediction = (
+                    not pushforward_trick
+                    or not self.training
+                    or step == max(batch.get_output_idxs())
+                )
+
+                if needs_full_prediction:
```
Comment (clessig) [inline src/weathergen/model/model.py]: We already call the forecast engine in l702. Don't we call it twice then in one iteration. 

Also, if we set `self.forecast_engine` to Identity if the number of blocks is 0 then we avoid the condition above.

File: src/weathergen/model/model.py [hunk 5]
```diff
@@ -695,15 +695,34 @@ def forward(self, model_params: ModelParams, batch: ModelBatch) -> ModelOutput:
...
         for step in batch.get_output_idxs():
             # apply forecasting engine (if present)
             if self.forecast_engine:
                 tokens = self.forecast_engine(tokens, step, coords=model_params.rope_coords)
 
-            # decoder predictions
-            output = self.predict_decoders(model_params, step, tokens, batch, output)
-            # latent predictions (raw and with SSL heads)
-            output = self.predict_latent(model_params, step, tokens, batch, output)
+                needs_full_prediction = (
```
Comment (clessig) [inline src/weathergen/model/model.py]: If you choose a bit more compact variable names then we can fit this in one line and it's more readable.

File: src/weathergen/model/model.py [hunk 6]
```diff
@@ -695,15 +695,34 @@ def forward(self, model_params: ModelParams, batch: ModelBatch) -> ModelOutput:
...
+                    not pushforward_trick
+                    or not self.training
+                    or step == max(batch.get_output_idxs())
```
Comment (clessig) [inline src/weathergen/model/model.py]: You use this to determine the number of forecast steps?

File: src/weathergen/model/model.py (hunk 5 above)
Comment (clessig) [reply src/weathergen/model/model.py]: Eg. `pushforward` instead of `pushforward_trick`

File: src/weathergen/model/model.py (hunk 6 above)
Comment (SavvasMel) [reply src/weathergen/model/model.py]: I use this to determine the last step for which we will take gradients.

File: src/weathergen/model/model.py (hunk 4 above)
Comment (SavvasMel) [reply src/weathergen/model/model.py]: 1) Sorry the one in l702 must have been introduced due to wrong merge, I will delete it.

2) Many thanks for the heads-up, I will correct this too.

Comment (SavvasMel) [reply src/weathergen/model/model.py]: @clessig just to make sure that I understand, if blocks are not >0 then the forecasting engine turns to None. Should I introduce an identity function instead in this case?

Comment (clessig) [reply src/weathergen/model/model.py]: Yes exactly

File: src/weathergen/model/engines.py [hunk 7]
```diff
@@ -390,6 +390,17 @@ def forward(self, tokens, coords=None):
         return tokens
 
 
+class IdentityEngine(torch.nn.Module):
+    """Identity engine that passes tokens through unchanged."""
+
+    def __init__(self):
+        super().__init__()
+        self.fe_blocks = torch.nn.ModuleList()
```
Comment (clessig) [inline src/weathergen/model/engines.py:537]: `self.blocks = ..`. Also do we need this as all? If anything I would expect `self.blocks = torch.nn.Identity()` but since we implement forward it might not be needed

File: src/weathergen/model/model.py [hunk 8]
```diff
@@ -697,16 +697,20 @@ def forward(self, model_params: ModelParams, batch: ModelBatch) -> ModelOutput:
         # collapse along input step dimension
         tokens = tokens.reshape(shape).sum(axis=1)
 
+        # Allow for pushforward trick
+        p_fwd = self.cf.get("pushforward_trick", False)
```
Comment (clessig) [inline src/weathergen/model/model.py]: I would expect that the push-forward trick config is part of forecast, i.e.

```
training_config : 

  .... 

  forecast: 
    num_steps: 3
    push_forward: True
```

**Review by clessig — COMMENTED** [review]
Essentially ready to merge, just some minor details.

File: src/weathergen/model/engines.py (hunk 7 above)
Comment (SavvasMel) [reply src/weathergen/model/engines.py:537]: I would like to keep it consistent:

```
class ForecastingEngine(torch.nn.Module):
    name: "ForecastingEngine"

    def __init__(self, cf: Config, mode_cfg, num_healpix_cells: int, dim_aux: int = None) -> None:
        """
        Initialize the ForecastingEngine with the configuration.

        :param cf: Configuration object containing parameters for the engine.
        :param num_healpix_cells: Number of healpix cells used for local queries.
        """
        super(ForecastingEngine, self).__init__()
        self.cf = cf
        self.num_healpix_cells = num_healpix_cells
        self.fe_blocks = torch.nn.ModuleList()        <-------------------
```

File: src/weathergen/model/model.py (hunk 8 above)
Comment (SavvasMel) [reply src/weathergen/model/model.py]: Done.

File: src/weathergen/model/engines.py (hunk 7 above)
Comment (SavvasMel) [reply src/weathergen/model/engines.py:537]: Also it allows self.fe_blocks to be iterable.

Comment (MatKbauer) [reply src/weathergen/model/engines.py:537]: Looks clean to me, but also wonder whether we need it. Are there cases where we want to iterate over a non-existing forecast engine?

File: src/weathergen/model/model.py [hunk 9]
```diff
@@ -704,16 +704,20 @@ def forward(self, model_params: ModelParams, batch: ModelBatch) -> ModelOutput:
...
+            all_grad = not p_fwd or not self.training or step == max(batch.get_output_idxs())
+
+            if all_grad:
+                tokens = self.forecast_engine(tokens, step, model_params.rope_coords)
+                output = self.predict_latent(model_params, step, tokens, batch, output)
+                output = self.predict_decoders(model_params, step, tokens, batch, output)
+            else:
+                # Pushforward mode: advance tokens without grad
+                with torch.no_grad():
+                    tokens = self.forecast_engine(tokens, step, model_params.rope_coords)
```
Comment (MatKbauer) [inline src/weathergen/model/model.py]:
```suggestion
            if self.forecast_engine:
                without_grad = not p_fwd or not self.training or step == max(batch.get_output_idxs())
                if without_grad:
                    # Pushforward mode: advance tokens without grad
                    with torch.no_grad():
                        tokens = self.forecast_engine(tokens, step, coords=model_params.rope_coords)
                        continue  # no decoding in pushforward mode
                tokens = self.forecast_engine(tokens, step, coords=model_params.rope_coords)

            # decoder predictions
            output = self.predict_decoders(model_params, step, tokens, batch, output)
            # latent predictions (raw and with SSL heads)
            output = self.predict_latent(model_params, step, tokens, batch, output)
```

Comment (MatKbauer) [reply src/weathergen/model/model.py]: I'd suggest to leave the forecast loop unchanged and call a `continue` statement in pushforward mode to not decode.

**Review by MatKbauer — CHANGES_REQUESTED** [review]
Pushforward logic is clean, but I advocate for keeping the forecast loop unchanged; see suggestion in the `forward()` function.

Also, I think we won't need the `IdentityEngine` and suggest to revert to the former code structure. With the suggested implementation of the pushforward logic, we would not need the explicit `IdentityEngine`.

File: src/weathergen/model/engines.py (hunk 7 above)
Comment (SavvasMel) [reply src/weathergen/model/engines.py:537]: It might be needed in the case we want to export the latent vectors even if FE is not present?

File: src/weathergen/model/model.py (hunk 9 above)
Comment (SavvasMel) [reply src/weathergen/model/model.py]: Since it has been decided to include an identity engine, the condition statement does not seem necessary. Also for the steps where we do not take gradients the output variable has automatically distributed empty lists for those steps in the initialization. I think the current state is more clean.

Any other suggestions/thoughts @clessig @MatKbauer ?

Comment (MatKbauer) [reply src/weathergen/model/model.py]: Agree on the identity engine. Concerning the order of calling the forecast engine, though, I found it more intuitive to leave the original order of `forecast_engine`, `predict_decoders`, `predict_latent` and add the push-forward mode before. Such as

```python
            for step in batch.get_output_idxs():

                without_grad = not p_fwd or not self.training or step == max(batch.get_output_idxs())
                if without_grad:
                    # Pushforward mode: advance tokens without grad
                    with torch.no_grad():
                        tokens = self.forecast_engine(tokens, step, coords=model_params.rope_coords)
                        continue  # no decoding in pushforward mode

                # Original code follows here, i.e.,
                tokens = self.forecast_engine(tokens, step, coords=model_params.rope_coords)
                # decoder predictions
                output = self.predict_decoders(model_params, step, tokens, batch, output)
                # latent predictions (raw and with SSL heads)
                output = self.predict_latent(model_params, step, tokens, batch, output)
```

Comment (MatKbauer) [reply src/weathergen/model/model.py]: This flow seems more approachable and readable to me.

Comment (SavvasMel) [reply src/weathergen/model/model.py]: Ok I will do the change, in that case, it should be:

`without_grad = p_fwd and self.training and step != max(batch.get_output_idxs())`

File: src/weathergen/model/model.py [hunk 10]
```diff
@@ -704,15 +704,20 @@ def forward(self, model_params: ModelParams, batch: ModelBatch) -> ModelOutput:
...
+        p_fwd = self.cf.training_config.get("forecast", {}).get("pushforward", False)
         # roll-out in latent space, iterate and generate output over requested output steps
         for step in batch.get_output_idxs():
-            # apply forecasting engine (if present)
-            if self.forecast_engine:
-                tokens = self.forecast_engine(tokens, step, coords=model_params.rope_coords)
+            without_grad = p_fwd and self.training and step != max(batch.get_output_idxs())
+            if without_grad:
+                # Pushforward mode: advance tokens without grad
+                # Pushforward mode: no decoding in this mode
```
Comment (MatKbauer) [inline src/weathergen/model/model.py]: Let's reduce this to a single comment. Something like `# Pushforward mode: advance tokens without grad; no decoding in this mode`

File: src/weathergen/model/model.py [hunk 11]
```diff
@@ -704,15 +704,20 @@ def forward(self, model_params: ModelParams, batch: ModelBatch) -> ModelOutput:
...
+                with torch.no_grad():
+                    tokens = self.forecast_engine(tokens, step, model_params.rope_coords)
+                    continue
 
-            # decoder predictions
+            tokens = self.forecast_engine(tokens, step, model_params.rope_coords)
             output = self.predict_decoders(model_params, step, tokens, batch, output)
-            # latent predictions (raw and with SSL heads)
             output = self.predict_latent(model_params, step, tokens, batch, output)
```
Comment (MatKbauer) [inline src/weathergen/model/model.py:721]: I find the two comments for the decoder (physical) and latent predictions useful. Can we keep them in?

**Review by MatKbauer — APPROVED** [review]
This looks great, thanks, Savvas! If we can still iterate once more over the comments in the forecast function, I'd be grateful.
